python/ray/util/client/server/dataservicer.py

```diff
@@ -34,7 +35,33 @@ def Datapath(self, request_iterator, context):
             return
         logger.debug(f"New data connection from client {client_id}: ")
         try:
-            for req in request_iterator:
+            request_queue = queue.Queue()
+
+            def fill_queue() -> None:
+                """
+                Pushes incoming requests to a shared request_queue.
+                """
+                try:
+                    for req in request_iterator:
+                        request_queue.put(req)
+                except grpc.RpcError as e:
+                    logger.debug("closing dataservicer reader thread "
+                                 f"grpc error reading request_iterator: {e}")
+                finally:
+                    request_queue.put(None)
+
+            queue_filler_thread = Thread(target=fill_queue, daemon=True)
```

Comment:
@ijrsvt 
Check out thread resource exhaustion like in [logservicer.py](https://github.com/ray-project/ray/blob/57c0bd9912caffc4cbb8a895784242663031a156/python/ray/util/client/server/logservicer.py#L93) and implement similar logic here.

This may change the factor of `CLIENT_SERVER_MAX_THREADS` to be `1/3`, but probably not.